study/src/org/labkey/study/model/DatasetDomainKind.java

```diff
@@ -435,6 +443,231 @@ else if (categoryId != null)
+            throw new IllegalArgumentException("\"Additional Key Column name \"" + keyPropertyName +"\" must be the name of a column.");
+
+        if (null != keyPropertyName && !useTimeKeyField)
+        {
+            String rangeURI = domain.getFieldByName(keyPropertyName).getRangeURI();
+            if (!(rangeURI.endsWith("int") || rangeURI.endsWith("double") || rangeURI.endsWith("string")))
+                throw new IllegalArgumentException("If Additional Key Column is managed, the column type must be numeric or text-based.");
+        }
+
+        // Other exception(s)
+
+        if (def != null &&  null != datasetId && def.getDatasetId() != datasetId && null != study.getDataset(datasetId))
+            throw new IllegalArgumentException("A Dataset already exists with the datasetId \"" + datasetId +"\".");
+    }
+
+    private void checkCanUpdate(DatasetDefinition def, Container container, User user,
+                                GWTDomain<? extends GWTPropertyDescriptor> original, GWTDomain<? extends GWTPropertyDescriptor> update)
+    {
+        if (null == def)
+            throw new IllegalArgumentException("Dataset not found.");
+
+        if (!def.canUpdateDefinition(user))
+            throw new IllegalArgumentException("Shared dataset can not be edited in this folder.");
+
+        if (null == PropertyService.get().getDomain(container, update.getDomainURI()))
+            throw new IllegalArgumentException("Domain not found: " + update.getDomainURI() + ".");
+
+        if (!def.getTypeURI().equals(original.getDomainURI()) || !def.getTypeURI().equals(update.getDomainURI()))
+            throw new IllegalArgumentException("Illegal Argument");
+    }
+
+    private ValidationException updateDomainDescriptor(GWTDomain<? extends GWTPropertyDescriptor> original, GWTDomain<? extends GWTPropertyDescriptor> update,
+                                                       Container container, User user, DatasetDefinition def)
+    {
+        ValidationException exception = new ValidationException();
+
+        try
+        {
+            List<String> allErrors = DomainUtil.updateDomainDescriptor(original, update, container, user).getAllErrors();
+            for (String str : allErrors) {
+                exception.addGlobalError(str);
+            }
+            return exception;
+        }
+        finally
+        {
+            StudyManager.getInstance().uncache(def);
+        }
+    }
+
+    private ValidationException updateDataset(DatasetDomainKindProperties datasetProperties, String domainURI, ValidationException exception,
+                                              StudyImpl study, Container container, User user, DatasetDefinition def)
+    {
+        try
+        {
+            // Update-case specific validation
+            if ( datasetProperties.isDemographicData() && !datasetProperties.isDemographicData() && !StudyManager.getInstance().isDataUniquePerParticipant(def))

```

Comment:
Looks like both of the first two parts of this if statement are using `datasetProperties`, I believe the second one should be using the `DatasetDefinition def` param:
```suggestion
            if ( datasetProperties.isDemographicData() && !def.isDemographicData() && !StudyManager.getInstance().isDataUniquePerParticipant(def))
```

Reply:
Also, I think we should move this validation check to the `checkCanUpdate()` method so that it lives with the other update dataset specific validation checks.